src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp

```diff
@@ -380,13 +380,14 @@ class Multiphase :
 
       real64 const dPhaseAmount_dP = dPorosity_dPressure * m_fluidPhaseSaturation( k, ip ) * m_fluidPhaseDensity( k, q, ip )
                                      + porosityNew * (m_dFluidPhaseSaturation_dPressure( k, ip ) * m_fluidPhaseDensity( k, q, ip )
-                                                      + m_fluidPhaseSaturation( k, ip ) * m_dFluidPhaseDensity_dPressure( k, q, ip ) );
+                                                      + m_fluidPhaseSaturation( k, ip ) * m_dFluidPhaseDensity( k, q, ip, Deriv::dP ) );
```

Comment:
In terms of memory layout, is having `Deriv::dP` as last component efficient?

Reply:
Here I just kept what we have now in the code (automatic permutation of the layout on the GPU). On the GPU, we permute the data layout as requested [here](https://github.com/GEOSX/GEOSX/blob/1dc46ce69fc126f0cd492be3f9f78661a434bb5e/src/coreComponents/constitutive/fluid/layouts.hpp#L56) to make the cell number last. This PR by Sergey discussed (and did) the implementation: https://github.com/GEOSX/GEOSX/pull/1463

Reply:
On CPU platforms, this places all derivatives in a single cell in contiguous memory, which is optimal because the access pattern for this data form a single core is basically sequential and values will be cached (with the caveat that we don't do anything to align cell data on cacheline size, so [false sharing](https://en.wikipedia.org/wiki/False_sharing) can occur; but that's a different story).

With CUDA, as @francoishamon mentioned, we permute the layout such that the unit-stride (contiguous) dimension is the one corresponding to cell index `k`, which results in perfectly coalesced access pattern optimal for GPUs.